docs(guide/component): fix for component controller unit-test example

mjwwit opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Docs update.

**What is the current behavior? (You can also link to an open issue here)**
Error in docs is described in #14426.

**Does this PR introduce a breaking change?**
No.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

mjwwit committed:
docs(guide/component): fix for component controller unit-test example (#14426)

Closes #14426

gkalpak commented on docs/content/guide/component.ngdoc:
```diff
@@ -233,9 +233,13 @@ it upwards to the heroList component, which updates the original data.
 
 <file name="heroDetail.js">
-  function HeroDetailController($scope, $element, $attrs) {
+  function HeroDetailController($scope) {
```
Do we need `$scope` here ? Did you leave it on purpose ?

mjwwit replied:
I made the absolute minimum number of changes to make the unit tests pass. In the unit tests the `$componentController` is provided with a `$scope` value, but not with `$element` or `$attrs`.
The `$scope` service is not being used in the controller though, so it might as well be removed.

If I'm not wrong, a PR was merged the other day removing the requirement to always provide `$scope` in `$componentController` locals, so it can be removed from there as well.
I can update the PR if you like.

gkalpak replied:
You are right. Please, do remove the unnecessary `$scope` and update the test to not provide one.

gkalpak commented on docs/content/guide/component.ngdoc:
```diff
@@ -233,9 +233,13 @@ it upwards to the heroList component, which updates the original data.
 
+    ctrl.hero = ctrl.hero || {};
+    ctrl.hero.name = ctrl.hero.name || undefined;
+    ctrl.hero.location = ctrl.hero.location || 'unknown';
```
Why did you make these changes ? They don't seem necessary.

mjwwit replied:
Those changes were tested in the unit tests. The tests were failing because the actual functionality was missing. I chose to update the component instead of removing the failing test. If needed, I can remove the test and the functionality.

gkalpak replied:
Really ? It's a shame we are not actually running these tests  :disappointed:

gkalpak replied:
TBH, I would rather remove this initialization and the corresponding test.
It doesn't add much value and doesn't seem like a good practice either. If nothing else, such initialization should be happening in an `$onInit` hook ?

WDYT ?

mjwwit replied:
I agree. I'll remove the test and the logic.

mjwwit commented:
I've removed the `$scope` references, and removed an outdated comment.

Isn't there some way you can get the unit tests to automatically run from docs using dgeni? It's already parsing the code snippets and putting them into files. Maybe it's possible to generate a karma config and run tests as well.

gkalpak commented:
We are running the e2e tests from the docs. Not sure if we are already doing the same for the guide.
Being able to run unit tests would be nice, but it won't be trivial I guess.

gkalpak commented on docs/content/guide/component.ngdoc:
```diff
@@ -464,8 +468,7 @@ describe('component: heroDetail', function() {
   it('should set the default values of the hero', function() {
-    // It's necessary to always pass the scope in the locals, so that the controller instance can be bound to it
-    component = $componentController('heroDetail', {$scope: scope});
+    component = $componentController('heroDetail', {});
```
Nitpick: I prefer to omit the unnecessary `locals` argument altogether (or pass `null` where we want a 3rd argument).

gkalpak commented:
Except for the fact that I am not a big fan of the initialization logic (see https://github.com/angular/angular.js/pull/14443/files#r60963112), it LGTM.
Feel free to also remove unnecessarily injected arguments to other controllers.
